server/src/main/java/org/elasticsearch/index/engine/PrunePostingsMergePolicy.java

```diff
@@ -49,22 +56,22 @@
  */
 final class PrunePostingsMergePolicy extends OneMergeWrappingMergePolicy {
 
-    PrunePostingsMergePolicy(MergePolicy in, String idField) {
+    PrunePostingsMergePolicy(MergePolicy in, String idField, Supplier<Query> retentionQuery) {
         super(in, toWrap -> new OneMerge(toWrap.segments) {
             @Override
             public CodecReader wrapForMerge(CodecReader reader) throws IOException {
                 CodecReader wrapped = toWrap.wrapForMerge(reader);
-                return wrapReader(wrapped, idField);
+                return wrapReader(wrapped, idField, retentionQuery);
             }
         });
     }
 
-    private static CodecReader wrapReader(CodecReader reader, String idField) {
-        Bits liveDocs = reader.getLiveDocs();
-        if (liveDocs == null) {
+    private static CodecReader wrapReader(CodecReader reader, String idField, Supplier<Query> retentionQuery) throws IOException {
+        if (reader.getLiveDocs() == null) {
             return reader; // no deleted docs - we are good!
         }
-        final boolean fullyDeletedSegment = reader.numDocs() == 0;
+        final Bits liveDocs = applyRetentionQuery(reader, retentionQuery.get());
```

Comment:
since we now use Lucene#union we can only consume it once. Should we move it below [here](https://github.com/elastic/elasticsearch/pull/43051/files#diff-3062b2c588c9f8342be195b47c9b8138R102)?

Reply:
now that I think about it we have to iterate this for every term so how does this work at all since we can only consume it once? I think we need to turn this into a bitset either way? We shoudl likely just use `BitSet#of` for this.